src/formatting/closures.rs

```diff
@@ -57,7 +57,14 @@ pub(crate) fn rewrite_closure(
 
         let result = match fn_decl.output {
             ast::FnRetTy::Default(_) if !context.inside_macro() => {
-                try_rewrite_without_block(body, &prefix, capture, context, shape, body_shape)
+                // Try-without-block only if configured and original code is not a block (#4394)
+                if !context.config.preserve_closure_block_wrapping()
+                    || !context.snippet(body.span).trim().starts_with('{')
+                {
```

Comment:
Thanks for the update with the config option! Bit of a nit, but I feel like we could write this in a way that's a bit more succinct and easy to read.

What would you think about moving these checks to the arm guard, or even creating an internal function (or closure :smile:) just above this match expression (perhaps something like `try_without_block` or `can_try_without_block`), which would contain the logic checks, and then just using that fn/closure call as the guard on the match arm?

Reply:
I did the changes (hopefully) as requested.  The function is called `can_try_rewrite_without_block()`.